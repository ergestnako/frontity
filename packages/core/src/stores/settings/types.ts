type State = {
  frontity: {
    name?: string;
    mode?: string;
  };
};

type Store = {
  state: {
    settings: State;
  };
};

type StoreWithSettings = Store & {
  settings: any;
};

export type Settings = {
  state: State;
  onInitialize?: (store: Store) => void;
  actions?: {
    addSettings: (_, store: StoreWithSettings) => void;
  };
};
