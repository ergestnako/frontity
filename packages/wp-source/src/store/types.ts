import { IConfig, IContext, IAction } from 'overmind';
import { namespaced } from 'overmind/config';
import * as source from '.';

const config = namespaced({ source });

export type Config = IConfig<typeof config>;
export type Context = IContext<Config>;
export type Action<Input = void> = IAction<Config, Input>;


// Action types
export type Fetch = Action<{
  name: string;
  page?: number;
}>;

type Payload = {
  name: string;
  params: { [param: string]: string };
  page: number;
};

export type Handler = (context: Context, payload: Payload) => Promise<void>;

export type Register = Action<{
  pattern: string | RegExp;
  handler: Handler;
}>;

// Effects types

export type IsMatch = (name: string, pattern: string) => boolean;
export type Exec = (name: string, pattern: string) => { [param: string]: string }