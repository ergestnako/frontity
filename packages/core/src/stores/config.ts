import { Namespace } from "@frontity/types/namespace";
import { namespaced } from "overmind/config";
import { Settings } from "./settings/types";

export default ({
  namespaces,
  settings
}: {
  namespaces: { [key: string]: Namespace };
  settings: Settings;
}) =>
  namespaced(
    Object.entries(namespaces).reduce(
      (config, [namespace, module]) => {
        if (module.Store) config[namespace] = module.Store;
        return config;
      },
      { settings }
    )
  );
