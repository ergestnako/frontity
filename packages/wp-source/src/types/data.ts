export type DataMap = {
  [name: string]: Data;
};

export type Data =
  | DataFetching
  | DataArchive
  | DataTaxonomy
  | DataCategory
  | DataTag
  | DataAuthor
  | DataPostTypeArchive
  | DataPostArchive
  | DataDate
  | DataPostType
  | DataPost
  | DataPage
  | DataAttachment
  | Data404;

// DATA FETCHING

export type DataFetching = {
  isFetching: boolean;
  isReady?: boolean;
};

// ARCHIVES

export type DataArchive = DataFetching & {
  isArchive: true;
  page: {
    type: string;
    id: number;
    link: string;
  }[][];
};

export type DataTaxonomy = DataArchive & {
  type: string;
  id: number;
  isTaxonomy: true;
};

export type DataCategory = DataTaxonomy & {
  isCategory: true;
};

export type DataTag = DataTaxonomy & {
  isTag: true;
};

export type DataAuthor = DataArchive & {
  id: number;
  isAuthor: true;
};

export type DataPostTypeArchive = DataFetching & {
  type: string;
  isPostTypeArchive: true;
};

export type DataPostArchive = DataPostTypeArchive & {
  isPostArchive: true;
  isHome: boolean;
  isFrontPage?: boolean;
};

export type DataDate = DataFetching & {
  date: string;
  isDate: true;
};

// POST TYPES

export type DataPostType = DataFetching & {
  type: string;
  id: number;
  isPostType: true;
};

export type DataPost = DataPostType & {
  isPost: true;
};

export type DataPage = DataPostType & {
  isPage: true;
  isFrontPage?: boolean;
};

export type DataAttachment = DataPostType & {
  isAttachment: true;
  isMedia: true;
};

// NOT FOUND (OR ERROR)

export type Data404 = DataFetching & {
  is404: true;
};
