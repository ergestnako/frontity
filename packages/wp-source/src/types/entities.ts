// ENTITIES

export type Entity = Single | Attachment | Taxonomy | Author;

export type Href = {
  href: string;
};

export type Embeddable = {
  embeddable: boolean;
};

export type Term = Href &
  Embeddable & {
    taxonomy: string;
  };

export type Links = {
  collection?: Href[];
  self?: Href[];
  about?: Href[];
  curies?: (Href & {
    name?: string;
    templated?: boolean;
  })[];
  author?: (Href & Embeddable)[];
  replies?: (Href & Embeddable)[];
  "wp:post_type"?: Href[];
  "version-history"?: Href[];
  "wp:attachment"?: Href[];
  "wp:contentmedia"?: (Href & Embeddable)[];
  "wp:featuredmedia"?: (Href & Embeddable)[];
  "wp:term"?: Term[];
};

export type Author = {
  type: "author";
  id: number;
  name: string;
  slug: string;
  description: string;
  link: string;
  meta: any;
  url: string;
  avatar_urls: {
    "24": string;
    "48": string;
    "96": string;
  };
  _links?: Links;
};

export type Taxonomy = {
  _links: Links;
  count: number;
  description: string;
  id: number;
  link: string;
  meta: any;
  name: string;
  parent: number;
  slug: string;
  taxonomy: string;
  type: string;
  yoast_meta?: {
    title: string;
  };
};

export type Embedded = {
  author: Author[];
  "wp:featuredmedia": Attachment[];
  "wp:contentmedia": Attachment[][];
  "wp:term": Taxonomy[];
};

export type Size = {
  file: string;
  height: number;
  mime_type: string; // TODO - set real mime/types
  source_url: string;
  width: number;
};

export type Attachment = {
  _embedded?: Embedded;
  _links: Links;
  alt_text: string;
  author: number;
  caption: string;
  comment_status: "open" | "closed";
  date: string;
  date_gmt: string;
  description: string;
  guid: {
    rendered: string;
  };
  id: number;
  link: string;
  media_details: {
    file: string;
    height: number;
    image_meta: {
      aperture: string;
      camera: string;
      caption: string;
      copyright: string;
      created_timestamp: string;
      credit: string;
      focal_length: string;
      iso: string;
      keywords: [];
      orientation: string;
      shutter_speed: string;
      title: string;
    };
    sizes: {
      full: Size;
      medium: Size;
      thumbnail: Size;
    };
    width: number;
  };
  media_type: string;
  meta: any;
  mime_type: string;
  modified: string;
  modified_gmt: string;
  ping_status: "closed" | "open";
  post: number;
  slug: string;
  source_url: string;
  title: {
    rendered: string;
  };
  type: "attachment";
  yoast_meta: {
    title: string;
  };
};

export type Single = {
  type: string;
  id: number;
  link: string;
  author: number;
  _links: Links;
  categories: number[];
  comment_status: "open" | "closed";
  content: {
    protected: boolean;
    rendered: string;
  };
  content_media: number[];
  date: string;
  date_gmt: string;
  excerpt: {
    protected: false;
    rendered: string;
  };
  featured_media: number;
  guid: {
    rendered: string;
  };
  format: string;
  meta: any;
  modified: string;
  modified_gmt: string;
  ping_status: "open" | "closed";
  slug: string;
  sticky: boolean;
  tags: number[];
  taxonomiesMap: {
    category: number[];
    tag: number[];
  };
  title: {
    rendered: string;
  };
  _embedded: Embedded;
  [other: string]: any; // is this necessary?
};

export type TaxonomyMap = {
  [name: string]: Taxonomy;
};

export type SingleMap = {
  [name: string]: Single;
};

export type AttachmentMap = {
  [name: string]: Attachment;
};

export type AuthorMap = {
  [name: string]: Author;
};