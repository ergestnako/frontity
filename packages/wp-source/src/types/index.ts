export * from "./data";
export * from "./store";
export * from "./entities";
