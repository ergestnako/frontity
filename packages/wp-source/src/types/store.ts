import { IConfig, IContext, IAction, IOnInitialize } from "overmind";
// import { namespaced } from "overmind/config";
import * as source from "../store";

const settings = {
  actions: {},
  effects: {},
  state: {
    packages: {
      source: {} as Settings
    }
  }
};

// const config = namespaced({ source, settings })

// type TConfig = typeof config;

// export type Config = IConfig<TConfig>;

export type Config = IConfig<{
  state: {
    source: typeof source.state;
    settings: typeof settings.state;
  };
  actions: {
    source: typeof source.actions;
    settings: typeof settings.actions;
  };
  effects: {
    source: typeof source.effects;
    settings: typeof settings.effects;
  };
}>;

export type Context = IContext<Config>;
export type Action<Input = void> = IAction<Config, Input>;

export type Handler = (
  context: Context,
  payload: {
    name: string;
    params: { [param: string]: any };
    page: number;
  }
) => Promise<void>;

// export type OnInitialize = IOnInitialize<Config>;

export type Settings = {
  apiUrl: string;
  isCom: boolean;
};