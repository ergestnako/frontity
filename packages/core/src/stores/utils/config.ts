import { Namespace } from "@frontity/types/namespace";
import { namespaced } from "overmind/config";

export default ({ namespaces }: { namespaces: { [key: string]: Namespace } }) =>
  namespaced(
    Object.entries(namespaces).reduce((config, [namespace, module]) => {
      if (module.Store) config[namespace] = module.Store;
      return config;
    }, {})
  );
