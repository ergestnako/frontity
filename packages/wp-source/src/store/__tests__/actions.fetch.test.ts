import { createOvermindMock } from "overmind";
import { namespaced } from "overmind/config";
import { actions, state, effects } from "..";
import { Handler, DataCategory } from "../../types";

const mockStore = () => {
  const config = namespaced({ source: { actions, state, effects } });

  const resolver = {
    match: (): {
      handler: Handler;
      params: { [param: string]: any };
    } => ({
      handler: async (ctx, { name, page }) => {
        const nature: DataCategory = {
          type: "category",
          id: 12,
          isArchive: true,
          isTaxonomy: true,
          isCategory: true,
          isFetching: true, 
          page: []
        };

        nature.page[page] = [
          { type: "post", id: 60, link: "/the-beauties-of-gullfoss" }
        ];

        Object.assign(ctx.state.source.data[name], nature);
      },
      params: {}
    })
  };

  return createOvermindMock(config, { source: { resolver } });
};

// return if the data that it's about to be fetched already exists
// init data

describe("actions â€º fetch", () => {
  test.only("adds data if that name doesn't exist yet", async () => {
    const name = "/category/nature";

    // init store
    const store = mockStore();

    // data[name] doesn't exist yet
    expect(store.state.source.data[name]).not.toBeDefined();

    // just after execute 'fetch', data[name] should exist
    const fetchPromise = store.actions.source.fetch({ name });
    expect(store.state.source.data[name]).toBeDefined();
    expect(store.state.source.data[name].isFetching).toBe(true);

    const mutations = await fetchPromise;
    expect(mutations).toMatchSnapshot();
  });

  test("doesn't do anything if that name exists", async () => {
    const name = "/category/nature";

    // init store
    const match = jest.fn(() => ({
      handler: async () => {},
      params: {}
    }));
    const store = mockStore();

    // add data[name] to the store
    await store.actions.source.fetch({ name });
    expect(match).toHaveBeenCalled();
    expect(store.state.source.data[name].isFetching).toBe(false);

    // fetch again the same name
    const fetchPromise = store.actions.source.fetch({ name });

    // just after execute nothing must have happened
    expect(match).toHaveBeenCalledTimes(1);
    expect(store.state.source.data[name].isFetching).toBe(false);

    // when fetch has finished nothing must have happened
    await fetchPromise;
    expect(match).toHaveBeenCalledTimes(1);
    expect(store.state.source.data[name].isFetching).toBe(false);
  });

  test("doesn't do anything if fetch is executed inside handler", async () => {
    const name = "/category/nature";

    // init store
    // Change 'resolver.match' so it calls 'source.fetch' action
    const match = jest.fn(() => {
      store.actions.source.fetch({ name });
      return Promise.resolve();
    });
    const store = mockStore();

    // spy on source.fetch
    jest.spyOn(store.actions.source, "fetch");

    /// just after execute 'fetch', data[name] should exist
    const fetchPromise = store.actions.source.fetch({ name });
    expect(store.state.source.data[name]).toBeDefined();
    expect(store.state.source.data[name].isFetching).toBe(true);

    // match function should have been called
    expect(match).toHaveBeenCalled();

    // when fetch has finished
    // 1. isFetching should be false
    // 2. fetch should have been called 2 times
    // 3. resolver.match should have been called once
    await fetchPromise;
    expect(store.state.source.data[name].isFetching).toBe(false);
    expect(store.actions.source.fetch).toHaveBeenCalledTimes(2);
    expect(match).toHaveBeenCalledTimes(1);
  });
});
