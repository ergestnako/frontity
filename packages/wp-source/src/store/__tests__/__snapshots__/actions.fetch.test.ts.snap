// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions â€º fetch adds data if that name doesn't exist yet 1`] = `
Array [
  Object {
    "args": Array [
      Object {
        "id": 12,
        "isArchive": true,
        "isCategory": true,
        "isFetching": false,
        "isTaxonomy": true,
        "page": Array [],
        "type": "category",
      },
    ],
    "method": "set",
    "path": "source.data./category/nature",
  },
  Object {
    "args": Array [
      "category",
    ],
    "method": "set",
    "path": "source.data./category/nature.type",
  },
  Object {
    "args": Array [
      12,
    ],
    "method": "set",
    "path": "source.data./category/nature.id",
  },
  Object {
    "args": Array [
      true,
    ],
    "method": "set",
    "path": "source.data./category/nature.isArchive",
  },
  Object {
    "args": Array [
      true,
    ],
    "method": "set",
    "path": "source.data./category/nature.isTaxonomy",
  },
  Object {
    "args": Array [
      true,
    ],
    "method": "set",
    "path": "source.data./category/nature.isCategory",
  },
  Object {
    "args": Array [
      true,
    ],
    "method": "set",
    "path": "source.data./category/nature.isFetching",
  },
  Object {
    "args": Array [
      Array [],
    ],
    "method": "set",
    "path": "source.data./category/nature.page",
  },
]
`;
